block/manager.go

```diff
@@ -678,7 +678,61 @@ func (m *Manager) createBlock(ctx context.Context, height uint64, lastSignature
 	return m.execCreateBlock(ctx, height, lastSignature, lastHeaderHash, m.lastState, batchData)
 }
 
-func (m *Manager) execCreateBlock(ctx context.Context, height uint64, lastSignature *types.Signature, lastHeaderHash types.Hash, lastState types.State, batchData *BatchData) (*types.SignedHeader, *types.Data, error) {
+func (m *Manager) applyBlock(ctx context.Context, header *types.SignedHeader, data *types.Data) (types.State, error) {
+	m.lastStateMtx.RLock()
+	defer m.lastStateMtx.RUnlock()
+	return m.execApplyBlock(ctx, m.lastState, header, data)
+}
+
+func (m *Manager) Validate(ctx context.Context, header *types.SignedHeader, data *types.Data) error {
+	m.lastStateMtx.RLock()
+	defer m.lastStateMtx.RUnlock()
+	return m.execValidate(m.lastState, header, data)
+}
+
+// execValidate validates a pair of header and data against the last state
+func (m *Manager) execValidate(lastState types.State, header *types.SignedHeader, data *types.Data) error {
+	// Validate the basic structure of the header
+	if err := header.ValidateBasic(); err != nil {
+		return fmt.Errorf("invalid header: %w", err)
+	}
+
+	// Validate the header against the data
+	if err := types.Validate(header, data); err != nil {
+		return fmt.Errorf("validation failed: %w", err)
+	}
+
+	// Ensure the header's Chain ID matches the expected state
+	if header.ChainID() != lastState.ChainID {
+		return fmt.Errorf("chain ID mismatch: expected %s, got %s", lastState.ChainID, header.ChainID())
+	}
+
+	// Check that the header's height is the expected next height
+	expectedHeight := lastState.LastBlockHeight + 1
+	if header.Height() != expectedHeight {
+		return fmt.Errorf("invalid height: expected %d, got %d", expectedHeight, header.Height())
+	}
+
+	// // Verify that the header's timestamp is strictly greater than the last block's time
```

Comment:
https://github.com/rollkit/rollkit/issues/2250 - marked as high priority to add it back